File: DMOJ Java/ds4_binary_search_tree_test/DS4_Splay.java
package ds4_binary_search_tree_test;

import java.io.BufferedReader;
import java.io.FileNotFoundException;
import java.io.FileReader;
import java.io.IOException;
import java.io.InputStream;
import java.io.InputStreamReader;
import java.io.OutputStreamWriter;
import java.io.PrintWriter;
import java.util.Iterator;
import java.util.NoSuchElementException;
import java.util.StringTokenizer;

public class DS4_Splay {

    private static DS4_Splay o = new DS4_Splay();
    
    public class Reader {
        private BufferedReader in;
        private StringTokenizer st;
        
        public Reader(InputStream inputStream) {
            in = new BufferedReader(new InputStreamReader(inputStream));
        } // Reader InputStream constructor
        
        public Reader(String fileName) throws FileNotFoundException {
            in = new BufferedReader(new FileReader(fileName));
        } // Reader String constructor

        public String next() throws IOException {
            while (st == null || !st.hasMoreTokens()) {
                st = new StringTokenizer(in.readLine().trim());
            } // while
            return st.nextToken();
        } // next method
        
        public long nextLong() throws IOException {
            return Long.parseLong(next());
        } // nextLong method
        
        public int nextInt() throws IOException {
            return Integer.parseInt(next());
        } // nextInt method
        
        public double nextDouble() throws IOException {
            return Double.parseDouble(next());
        } // nextDouble method
        
        public String nextLine() throws IOException {
            return in.readLine().trim();
        } // nextLine method
    } // Reader class
    
    public class Queue<Item> implements Iterable<Item> {
        private Node<Item> first;    // beginning of queue
        private Node<Item> last;     // end of queue
        private int n;               // number of elements on queue

        // helper linked list class
        private class Node<Item> {
            private Item item;
            private Node<Item> next;
        }

        /**
         * Initializes an empty queue.
         */
        public Queue() {
            first = null;
            last  = null;
            n = 0;
        }

        /**
         * Returns true if this queue is empty.
         *
         * @return {@code true} if this queue is empty; {@code false} otherwise
         */
        public boolean isEmpty() {
            return first == null;
        }

        /**
         * Returns the number of items in this queue.
         *
         * @return the number of items in this queue
         */
        public int size() {
            return n;
        }

        /**
         * Returns the item least recently added to this queue.
         *
         * @return the item least recently added to this queue
         * @throws NoSuchElementException if this queue is empty
         */
        public Item peek() {
            if (isEmpty()) throw new NoSuchElementException("Queue underflow");
            return first.item;
        }

        /**
         * Adds the item to this queue.
         *
         * @param  item the item to add
         */
        public void enqueue(Item item) {
            Node<Item> oldlast = last;
            last = new Node<Item>();
            last.item = item;
            last.next = null;
            if (isEmpty()) first = last;
            else           oldlast.next = last;
            n++;
        }

        /**
         * Removes and returns the item on this queue that was least recently added.
         *
         * @return the item on this queue that was least recently added
         * @throws NoSuchElementException if this queue is empty
         */
        public Item dequeue() {
            if (isEmpty()) throw new NoSuchElementException("Queue underflow");
            Item item = first.item;
            first = first.next;
            n--;
            if (isEmpty()) last = null;   // to avoid loitering
            return item;
        }

        /**
         * Returns a string representation of this queue.
         *
         * @return the sequence of items in FIFO order, separated by spaces
         */
        public String toString() {
            StringBuilder s = new StringBuilder();
            for (Item item : this) {
                s.append(item);
                s.append(' ');
            }
            return s.toString();
        } 

        /**
         * Returns an iterator that iterates over the items in this queue in FIFO order.
         *
         * @return an iterator that iterates over the items in this queue in FIFO order
         */
        public Iterator<Item> iterator()  {
            return new ListIterator<Item>(first);  
        }

        // an iterator, doesn't implement remove() since it's optional
        private class ListIterator<Item> implements Iterator<Item> {
            private Node<Item> current;

            public ListIterator(Node<Item> first) {
                current = first;
            }

            public boolean hasNext()  { return current != null;                     }
            public void remove()      { throw new UnsupportedOperationException();  }

            public Item next() {
                if (!hasNext()) throw new NoSuchElementException();
                Item item = current.item;
                current = current.next; 
                return item;
            }
        }
    }
    
    public class SplayTreeSet<Value extends Comparable<Value>>  {

        private Node root;   // root of the BST

        // BST helper node data type
        private class Node {
            private Value val;
            private Node left, right;   // left and right subtrees
            private int size;

            public Node(Value val, int size) {
                this.val = val;
                this.size = size;
            }
        }

        public boolean contains(Value val) {
            if (root == null) return false;
            root = splay(root, val);
            return val.compareTo(root.val) == 0;
        }
        
        /**
         * Returns the kth smallest value in the symbol table.
         * 
         * @param k the order statistic
         * @return the kth smallest value in the symbol table
         * @throws IllegalArgumentException unless {@code k} is between 0 and
         *             {@code size() -1 }
         */
        public Value select(int k) {
            if (k < 0 || k >= size()) throw new IllegalArgumentException("k is not in range 0-" + (size() - 1));
            return select(root, k).val;
        }
        
        // auxiliary method for select
        private Node select(Node x, int k) {
            if (x == null) return null;
            int t = size(x.left);
            if (t > k) return select(x.left, k);
            else if (t < k) return select(x.right, k - t - 1);
            return x;
        }
        
        /**
         * Returns the number of values in the symbol table strictly less than
         * {@code val}.
         * 
         * @param val the value
         * @return the number of values in the symbol table strictly less than
         *         {@code val}
         * @throws IllegalArgumentException if {@code val} is {@code null}
         */
        public int rank(Value val) {
            if (val == null) throw new IllegalArgumentException("argument to rank() is null");
            root = splay(root, val);
            return rank(root, val);
        }

        /**
         * Returns the number of values in the subtree less than val.
         * 
         * @param val the value
         * @param x the subtree
         * @return the number of values in the subtree less than val
         */
        private int rank(Node x, Value val) {
            if (x == null) return 0;
            if (val.compareTo(x.val) <= 0) return rank(x.left, val);
            else return 1 + size(x.left) + rank(x.right, val);
        }  

       /***************************************************************************
        *  Splay tree insertion.
        ***************************************************************************/
        public void add(Value val) {
            // splay value to root
            if (root == null) {
                root = new Node(val, 1);
                return;
            }
            root = splay(root, val);
            int cmp = val.compareTo(root.val);
            // Insert new node at root
            if (cmp < 0) {
                Node n = new Node(val, 1);
                n.left = root.left;
                n.right = root;
                root.left = null;
                update(root);
                root = n;
            }
            // Insert new node at root
            else {
                Node n = new Node(val, 1);
                n.right = root.right;
                n.left = root;
                root.right = null;
                update(root);
                root = n;
            }
            update(root);
        }
        
       /***************************************************************************
        *  Splay tree deletion.
        ***************************************************************************/
        /* This splays the value, then does a slightly modified Hibbard deletion on
         * the root (if it is the node to be deleted; if it is not, the value was 
         * not in the tree). The modification is that rather than swapping the
         * root (call it node A) with its successor, it's successor (call it Node B)
         * is moved to the root position by splaying for the deletion value in A's 
         * right subtree. Finally, A's right child is made the new root's right 
         * child.
         */
        public void remove(Value val) {
            if (root == null) return; // empty tree
            root = splay(root, val);
            int cmp = val.compareTo(root.val);
            if (cmp == 0) {
                if (root.left == null) {
                    root = root.right;
                } else {
                    Node x = root.right;
                    root = splay(root.left, val);
                    root.right = x;
                }
                update(root);
            }
            // else: it wasn't in the tree to remove
        }
        
        /**
         * Returns an iterator that iterates over the values in order
         * 
         * @return an iterator that iterates over the values in order
         */
        public Iterable<Value> values() {
            Queue<Value> order = new Queue<Value>();
            traverse(root, order);
            return order;
        }
        
        private void traverse(Node x, Queue<Value> order) {
            if (x == null) return;
            traverse(x.left, order);
            order.enqueue(x.val);
            traverse(x.right, order);
        }
        
        
       /***************************************************************************
        * Splay tree function.
        * **********************************************************************/
        // splay value in the tree rooted at Node h. If a node with that value exists,
        //   it is splayed to the root of the tree. If it does not, the last node
        //   along the search path for the value is splayed to the root.
        private Node splay(Node h, Value val) {
            if (h == null) return null;
            int cmp1 = val.compareTo(h.val);
            if (cmp1 < 0) {
                // key not in tree, so we're done
                if (h.left == null)  return h;
                int cmp2 = val.compareTo(h.left.val);
                if (cmp2 < 0) {
                    h.left.left = splay(h.left.left, val);
                    h = rotateRight(h);
                } else if (cmp2 > 0) {
                    h.left.right = splay(h.left.right, val);
                    if (h.left.right != null) {
                        h.left = rotateLeft(h.left);
                    }
                }
                if (h.left == null) return h;
                else return rotateRight(h);
            } else if (cmp1 > 0) { 
                // key not in tree, so we're done
                if (h.right == null) return h;
                int cmp2 = val.compareTo(h.right.val);
                if (cmp2 < 0) {
                    h.right.left = splay(h.right.left, val);
                    if (h.right.left != null) {
                        h.right = rotateRight(h.right);
                    }
                } else if (cmp2 > 0) {
                    h.right.right = splay(h.right.right, val);
                    h = rotateLeft(h);
                }
                if (h.right == null) return h;
                else return rotateLeft(h);
            } else {
                return h;
            }
        }


       /***************************************************************************
        *  Helper functions.
        ***************************************************************************/
        
        public int size() {
            return size(root);
        }
        
        private int size(Node x) {
            if (x == null) return 0;
            return x.size;
        }
        
        /**
         * Updates the size  of the subtree.
         *
         * @param x the subtree
         */
        private void update(Node x) {
            if (x == null) return;
            x.size = 1 + size(x.left) + size(x.right);
        }
        
        // right rotate
        private Node rotateRight(Node h) {
            Node x = h.left;
            h.left = x.right;
            x.right = h;
            update(h);
            update(x);
            return x;
        }

        // left rotate
        private Node rotateLeft(Node h) {
            Node x = h.right;
            h.right = x.left;
            x.left = h;
            update(h);
            update(x);
            return x;
        }
    }
    
    public static void main(String[] args) throws IOException {
        Reader in = o.new Reader(System.in);
        PrintWriter out = new PrintWriter(new OutputStreamWriter(System.out));
        int n = in.nextInt();
        int m = in.nextInt();
        SplayTreeSet<Integer> tree = o.new SplayTreeSet<Integer>();
        for (int i = 0; i < n; i++) {
            int x = in.nextInt();
            tree.add(x);
        }
        int lastAns = 0;
        for (int i = 0; i < m; i++) {
            String op = in.next();
            int x = in.nextInt() ^ lastAns;
            if (op.equals("I")) {
                tree.add(x);
            } else if (op.equals("R")) {
                tree.remove(x);
            } else if (op.equals("S")) {
                lastAns = tree.select(x - 1);
                out.println(lastAns);
            } else /*if (op.equals("L"))*/{
                lastAns = tree.contains(x) ? tree.rank(x) + 1 : -1;
                out.println(lastAns);
            }
        }
        for (int val: tree.values()) {
            out.print(val + " ");
        }
        out.println();
        out.close();
    }
}

File: DMOJ C++/ds4_binary_search_tree_test_splay/DS4_Splay.cpp
/*
 * DS4_Splay.cpp
 *
 *  Created on: Jul 20, 2017
 *      Author: Wesley Leung
 */

#include <bits/stdc++.h>

#define SHORT_INF 0x3f3f
#define INT_INF 0x3f3f3f3f
#define LL_INF 0x3f3f3f3f3f3f3f3f
#define pb push_back
#define mp make_pair
#define l(x) x << 1
#define r(x) x << 1 | 1
#define mid(x, y) x + (y - x) / 2
#define f first
#define s second
#define ri(x) scanf("%d", &x)
#define rll(x) scanf("%lld", &x)
#define rllu(x) scanf("%llu", &x)
#define rf(x) scanf("%f", &x)
#define rd(x) scanf("%lf", &x)
#define rc(x) scanf(" %c", &x)
#define rs(x) scanf("%s", x)
#define Fill(a, x) memset(a, x, sizeof(a))
#define randi(a, b) rand() % (b - a + 1) + a

using namespace std;

typedef long long ll;
typedef unsigned long long llu;
typedef pair<int, int> pii;
typedef pair<float, float> pff;
typedef pair<double, double> pdd;
typedef pair<ll, ll> pll;
typedef pair<llu, llu> pllu;
typedef map<int, int> mii;
typedef map<int, ll> mill;
typedef map<ll, int> mlli;
typedef unordered_map<int, int> umii;
typedef unordered_map<int, ll> umill;
typedef unordered_map<ll, int> umlli;

template <typename Value>
struct SplayTreeSet {
    // BST helper node data type
    struct Node {
        Value val;
        Node *left = nullptr, *right = nullptr;   // left and right subtrees
        int size;
        Node(Value val, int size) {
            this->val = val;
            this->size = size;
        }
    };

private:
    Node *root = nullptr;   // root of the BST

    int size(Node *&x) {
        if (x == nullptr) return 0;
        return x->size;
    }

    /**
     * Updates the size and height of the subtree.
     *
     * @param x the subtree
     */
    void update(Node *&x) {
        if (x == nullptr) return;
        x->size = 1 + size(x->left) + size(x->right);
    }

    /**
     * Rotates the given subtree to the right.
     *
     * @param x the subtree
     * @return the right rotated subtree
     */
    Node *rotateRight(Node *&x) {
        Node *y = x->left;
        x->left = y->right;
        y->right = x;
        update(x);
        update(y);
        return y;
    }

    /**
     * Rotates the given subtree to the left.
     *
     * @param x the subtree
     * @return the left rotated subtree
     */
    Node *rotateLeft(Node *&x) {
        Node *y = x->right;
        x->right = y->left;
        y->left = x;
        update(x);
        update(y);
        return y;
    }

    // auxiliary method for select
    Node *select(Node *&x, int k) {
        if (x == nullptr) return nullptr;
        int t = size(x->left);
        if (t > k) return select(x->left, k);
        else if (t < k) return select(x->right, k - t - 1);
        return x;
    }

    /**
     * Returns the number of values in the subtree less than val.
     *
     * @param val the value
     * @param x the subtree
     * @return the number of values in the subtree less than val
     */
    int getRank(Node *&x, Value val) {
        if (x == nullptr) return 0;
        if (val <= x->val) return getRank(x->left, val);
        else return 1 + size(x->left) + getRank(x->right, val);
    }

    void traverse(Node *&x, vector<Value> &order) {
        if (x == nullptr) return;
        traverse(x->left, order);
        order.push_back(x->val);
        traverse(x->right, order);
    }

    /***************************************************************************
     * Splay tree function.
     * **********************************************************************/
     // splay value in the tree rooted at Node x. If a node with that value exists,
     //   it is splayed to the root of the tree. If it does not, the last node
     //   along the search path for the value is splayed to the root.
     Node *splay(Node *&x, Value val) {
         if (x == nullptr) return nullptr;
         if (val < x->val) {
             // key not in tree, so we're done
             if (x->left == nullptr) return x;
             if (val < x->left->val) {
                 x->left->left = splay(x->left->left, val);
                 x = rotateRight(x);
             } else if (val > x->left->val) {
                 x->left->right = splay(x->left->right, val);
                 if (x->left->right != nullptr) x->left = rotateLeft(x->left);
             }
             if (x->left == nullptr) return x;
             else return rotateRight(x);
         } else if (val > x->val) {
             // key not in tree, so we're done
             if (x->right == nullptr) return x;
             if (val < x->right->val) {
                 x->right->left = splay(x->right->left, val);
                 if (x->right->left != nullptr) x->right = rotateRight(x->right);
             } else if (val > x->right->val) {
                 x->right->right = splay(x->right->right, val);
                 x = rotateLeft(x);
             }
             if (x->right == nullptr) return x;
             else return rotateLeft(x);
         } else {
             return x;
         }
     }

public:
    SplayTreeSet() {}

    int size() {
        return size(root);
    }

    bool contains(Value val) {
        if (root == nullptr) return false;
        root = splay(root, val);
        return val == root->val;
    }

    /**
     * Returns the kth smallest value in the symbol table.
     *
     * @param k the order statistic
     * @return the kth smallest value in the symbol table
     * @throws IllegalArgumentException unless {@code k} is between 0 and
     *             {@code size() -1 }
     */
    Value select(int k) {
        if (k < 0 || k >= size()) throw invalid_argument("k is not in range 0 to size");
        return select(root, k)->val;
    }

    /**
     * Returns the number of values in the symbol table strictly less than
     * {@code val}.
     *
     * @param val the value
     * @return the number of values in the symbol table strictly less than
     *         {@code val}
     */
    int getRank(Value val) {
        root = splay(root, val);
        return getRank(root, val);
    }

   /***************************************************************************
    *  Splay tree insertion.
    ***************************************************************************/
    void add(Value val) {
        // splay value to root
        if (root == nullptr) {
            root = new Node(val, 1);
            return;
        }
        root = splay(root, val);
        // Insert new node at root
        if (val < root->val) {
            Node *n = new Node(val, 1);
            n->left = root->left;
            n->right = root;
            root->left = nullptr;
            update(root);
            root = n;
        }
        // Insert new node at root
        else {
            Node *n = new Node(val, 1);
            n->right = root->right;
            n->left = root;
            root->right = nullptr;
            update(root);
            root = n;
        }
        update(root);
    }

   /***************************************************************************
    *  Splay tree deletion.
    ***************************************************************************/
    /* This splays the value, then does a slightly modified Hibbard deletion on
     * the root (if it is the node to be deleted; if it is not, the value was
     * not in the tree). The modification is that rather than swapping the
     * root (call it node A) with its successor, it's successor (call it Node B)
     * is moved to the root position by splaying for the deletion value in A's
     * right subtree. Finally, A's right child is made the new root's right
     * child.
     */
    void remove(Value val) {
        if (root == nullptr) return; // empty tree
        root = splay(root, val);
        if (val == root->val) {
            if (root->left == nullptr) {
                root = root->right;
            } else {
                Node *x = root->right;
                root = splay(root->left, val);
                root->right = x;
            }
            update(root);
        }
        // else: it wasn't in the tree to remove
    }

    /**
     * Returns an iterator that iterates over the values in order
     *
     * @return an iterator that iterates over the values in order
     */
    vector<Value> values() {
        vector<Value> order;
        traverse(root, order);
        return order;
    }
};

int N, M;
SplayTreeSet<int> tree;

int main() {
    ri(N);
    ri(M);
    for (int i = 0; i < N; i++) {
        int x;
        ri(x);
        tree.add(x);
    }
    int lastAns = 0;
    for (int i = 0; i < M; i++) {
        char op;
        int x;
        rc(op);
        ri(x);
        x = x ^ lastAns;
        if (op == 'I') {
            tree.add(x);
        } else if (op == 'R') {
            tree.remove(x);
        } else if (op == 'S') {
            lastAns = tree.select(x - 1);
            printf("%d\n", lastAns);
        } else if (op == 'L') {
            lastAns = tree.contains(x) ? tree.getRank(x) + 1 : -1;
            printf("%d\n", lastAns);
        } else {
            i--;
        }
    }
    for (int x: tree.values()) {
        printf("%d ", x);
    }
    return 0;
}
